File: projects/rpn2tex/migrations_archive/java-feature-by-feature-1/src/main/java/com/rpn2tex/Lexer.java
package com.rpn2tex;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * Tokenizes RPN input text.
 *
 * <p>The lexer scans input character by character, producing tokens for:
 * <ul>
 *   <li>Numbers (integers and decimals, including negative numbers)</li>
 *   <li>Operators (+, -, *, /)</li>
 *   <li>EOF marker</li>
 * </ul>
 *
 * <p>Whitespace is used as a delimiter and is otherwise ignored.
 *
 * <p>The '-' character is disambiguated using lookahead:
 * <ul>
 *   <li>If followed immediately by a digit: negative number (NUMBER token)</li>
 *   <li>Otherwise: subtraction operator (MINUS token)</li>
 * </ul>
 *
 * <p>Example usage:
 * <pre>
 *   Lexer lexer = new Lexer("5 3 -");
 *   List&lt;Token&gt; tokens = lexer.tokenize();
 *   // Returns: [Token(NUMBER, '5', 1:1), Token(NUMBER, '3', 1:3), Token(MINUS, '-', 1:5), Token(EOF, '', 1:6)]
 * </pre>
 */
public class Lexer {
    private final String text;
    private int pos;
    private int line;
    private int column;

    /**
     * Creates a new lexer for the given input text.
     *
     * @param text the RPN expression to tokenize
     * @throws NullPointerException if text is null
     */
    public Lexer(String text) {
        this.text = Objects.requireNonNull(text, "text must not be null");
        this.pos = 0;
        this.line = 1;
        this.column = 1;
    }

    /**
     * Tokenizes the entire input text.
     *
     * @return list of tokens, ending with EOF token
     * @throws LexerException if an invalid character is encountered
     */
    public List<Token> tokenize() throws LexerException {
        List<Token> tokens = new ArrayList<>();

        while (!atEnd()) {
            skipWhitespace();
            if (atEnd()) {
                break;
            }
            tokens.add(scanToken());
        }

        // Add EOF token
        tokens.add(new Token(TokenType.EOF, "", line, column));
        return tokens;
    }

    private boolean atEnd() {
        return pos >= text.length();
    }

    private char peek() {
        if (atEnd()) {
            return '\0';
        }
        return text.charAt(pos);
    }

    private char advance() {
        char ch = text.charAt(pos);
        pos++;
        if (ch == '\n') {
            line++;
            column = 1;
        } else {
            column++;
        }
        return ch;
    }

    private void skipWhitespace() {
        while (!atEnd() && isWhitespace(peek())) {
            advance();
        }
    }

    private boolean isWhitespace(char ch) {
        return ch == ' ' || ch == '\t' || ch == '\n' || ch == '\r';
    }

    private Token scanToken() throws LexerException {
        int startLine = line;
        int startColumn = column;
        char ch = peek();

        // Single-character operators
        if (ch == '+') {
            advance();
            return new Token(TokenType.PLUS, "+", startLine, startColumn);
        }

        if (ch == '*') {
            advance();
            return new Token(TokenType.MULT, "*", startLine, startColumn);
        }

        if (ch == '/') {
            advance();
            return new Token(TokenType.DIV, "/", startLine, startColumn);
        }

        // Handle minus: distinguish between negative number and subtraction operator
        if (ch == '-') {
            advance();
            // If followed immediately by a digit, it's a negative number
            if (!atEnd() && Character.isDigit(peek())) {
                return scanNumber("-", startLine, startColumn);
            }
            // Otherwise, it's a subtraction operator
            return new Token(TokenType.MINUS, "-", startLine, startColumn);
        }

        // Numbers
        if (Character.isDigit(ch)) {
            return scanNumber("", startLine, startColumn);
        }

        // Unknown character
        throw new LexerException(
            String.format("Unexpected character '%c'", ch),
            startLine,
            startColumn
        );
    }

    private Token scanNumber(String prefix, int startLine, int startColumn) {
        StringBuilder value = new StringBuilder(prefix);

        // Integer part
        while (!atEnd() && Character.isDigit(peek())) {
            value.append(advance());
        }

        // Decimal part (optional)
        if (!atEnd() && peek() == '.') {
            value.append(advance()); // consume '.'
            while (!atEnd() && Character.isDigit(peek())) {
                value.append(advance());
            }
        }

        return new Token(TokenType.NUMBER, value.toString(), startLine, startColumn);
    }
}

File: projects/rpn2tex/migrations_archive/rust-feature-by-feature-4/src/lexer.rs
//! Lexer for rpn2tex - converts text into tokens.
//!
//! This module tokenizes RPN (Reverse Polish Notation) expressions.

use std::iter::Peekable;
use std::str::Chars;

use crate::error::LexerError;
use crate::tokens::{Token, TokenType};

/// Tokenizes RPN input text.
///
/// The lexer scans input character by character, producing tokens for:
/// - Numbers (integers and decimals)
/// - EOF marker
///
/// Whitespace is used as a delimiter and is otherwise ignored.
///
/// # Examples
///
/// ```
/// use rpn2tex::lexer::Lexer;
///
/// let lexer = Lexer::new("5");
/// let tokens = lexer.tokenize().unwrap();
/// assert_eq!(tokens.len(), 2); // NUMBER and EOF
/// ```
pub struct Lexer<'a> {
    /// Iterator over characters in the input
    chars: Peekable<Chars<'a>>,
    /// Current line number (1-based)
    line: usize,
    /// Current column number (1-based)
    column: usize,
}

impl<'a> Lexer<'a> {
    /// Create a new lexer with input text.
    ///
    /// # Examples
    ///
    /// ```
    /// use rpn2tex::lexer::Lexer;
    ///
    /// let lexer = Lexer::new("5 3 +");
    /// ```
    #[must_use]
    pub fn new(text: &'a str) -> Self {
        Self {
            chars: text.chars().peekable(),
            line: 1,
            column: 1,
        }
    }

    /// Tokenize the entire input text.
    ///
    /// Returns a list of tokens, ending with an EOF token.
    ///
    /// # Errors
    ///
    /// Returns `LexerError` if an invalid character is encountered.
    ///
    /// # Examples
    ///
    /// ```
    /// use rpn2tex::lexer::Lexer;
    ///
    /// let lexer = Lexer::new("5");
    /// let tokens = lexer.tokenize().unwrap();
    /// ```
    pub fn tokenize(mut self) -> Result<Vec<Token>, LexerError> {
        let mut tokens = Vec::new();

        while self.peek().is_some() {
            self.skip_whitespace();
            if self.peek().is_none() {
                break;
            }
            tokens.push(self.scan_token()?);
        }

        // Add EOF token
        tokens.push(Token::new(
            TokenType::Eof,
            String::new(),
            self.line,
            self.column,
        ));
        Ok(tokens)
    }

    /// Look at current character without consuming it.
    fn peek(&mut self) -> Option<char> {
        self.chars.peek().copied()
    }

    /// Consume and return the current character.
    fn advance(&mut self) -> Option<char> {
        let ch = self.chars.next()?;
        if ch == '\n' {
            self.line += 1;
            self.column = 1;
        } else {
            self.column += 1;
        }
        Some(ch)
    }

    /// Skip over whitespace characters.
    fn skip_whitespace(&mut self) {
        while let Some(ch) = self.peek() {
            if ch.is_whitespace() {
                self.advance();
            } else {
                break;
            }
        }
    }

    /// Scan and return the next token.
    ///
    /// # Errors
    ///
    /// Returns `LexerError` if an invalid character is encountered.
    fn scan_token(&mut self) -> Result<Token, LexerError> {
        let start_line = self.line;
        let start_column = self.column;

        let Some(ch) = self.peek() else {
            return Err(LexerError::new(
                "Unexpected end of input".to_string(),
                start_line,
                start_column,
            ));
        };

        // Numbers
        if ch.is_ascii_digit() {
            return self.scan_number(String::new(), start_line, start_column);
        }

        // Handle negative numbers and subtraction operator
        if ch == '-' {
            self.advance();
            // Check if this is a negative number (digit follows immediately)
            if let Some(next_ch) = self.peek() {
                if next_ch.is_ascii_digit() {
                    return self.scan_number("-".to_string(), start_line, start_column);
                }
            }
            // Otherwise, it's a subtraction operator
            return Ok(Token::new(
                TokenType::Minus,
                "-".to_string(),
                start_line,
                start_column,
            ));
        }

        // Addition operator
        if ch == '+' {
            self.advance();
            return Ok(Token::new(
                TokenType::Plus,
                "+".to_string(),
                start_line,
                start_column,
            ));
        }

        // Multiplication operator
        if ch == '*' {
            self.advance();
            return Ok(Token::new(
                TokenType::Mult,
                "*".to_string(),
                start_line,
                start_column,
            ));
        }

        // Division operator
        if ch == '/' {
            self.advance();
            return Ok(Token::new(
                TokenType::Div,
                "/".to_string(),
                start_line,
                start_column,
            ));
        }

        // Unknown character
        Err(LexerError::new(
            format!("Unexpected character '{ch}'"),
            start_line,
            start_column,
        ))
    }

    /// Scan a numeric literal.
    fn scan_number(
        &mut self,
        prefix: String,
        start_line: usize,
        start_column: usize,
    ) -> Result<Token, LexerError> {
        let mut value = prefix;

        // Integer part
        while let Some(ch) = self.peek() {
            if ch.is_ascii_digit() {
                value.push(ch);
                self.advance();
            } else {
                break;
            }
        }

        // Decimal part (optional)
        if let Some('.') = self.peek() {
            value.push('.');
            self.advance();

            while let Some(ch) = self.peek() {
                if ch.is_ascii_digit() {
                    value.push(ch);
                    self.advance();
                } else {
                    break;
                }
            }
        }

        Ok(Token::new(
            TokenType::Number,
            value,
            start_line,
            start_column,
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_integer() {
        let lexer = Lexer::new("5");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens.len(), 2);
        assert_eq!(tokens[0].token_type, TokenType::Number);
        assert_eq!(tokens[0].value, "5");
    }

    #[test]
    fn test_decimal() {
        let lexer = Lexer::new("3.14");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens.len(), 2);
        assert_eq!(tokens[0].token_type, TokenType::Number);
        assert_eq!(tokens[0].value, "3.14");
    }

    #[test]
    fn test_division_operator() {
        let lexer = Lexer::new("10 2 /");
        let tokens = lexer.tokenize().unwrap();
        assert_eq!(tokens.len(), 4); // NUMBER, NUMBER, DIV, EOF
        assert_eq!(tokens[0].token_type, TokenType::Number);
        assert_eq!(tokens[0].value, "10");
        assert_eq!(tokens[1].token_type, TokenType::Number);
        assert_eq!(tokens[1].value, "2");
        assert_eq!(tokens[2].token_type, TokenType::Div);
        assert_eq!(tokens[2].value, "/");
        assert_eq!(tokens[3].token_type, TokenType::Eof);
    }
}
